qmetaobject/src/qrc.rs

```diff
@@ -21,39 +21,84 @@ Q_CORE_EXPORT bool qRegisterResourceData(int, const unsigned char *,
                                          const unsigned char *, const unsigned char *);
 }}
 
-/// Macro to embed files and made them available to the Qt resource system
+/// Embed files and made them available to the Qt resource system.
 ///
-/// ```ignore
+/// The macro accepts an identifier with optional preceding visibility modifier,
+/// and a comma-separated list of resources. Then macro generates a function
+/// with given name and visibility, which can be used to register all the
+/// resources.
+///
+/// # Input
+///
+/// The macro accepts an identifier with optional preceding visibility modifier
+/// and a comma-separated list of _resources_. Each _resource_ consists of a prefix
+/// path and a braced list of comma-separated _files_. Each _file_ is specified as
+/// a path with optional alias path after `as` keyword.
+///
+/// The paths are relative to the location in which cargo runs.
+///
+/// It does not matter if the prefix has leading '/' or not.
+///
+/// # Output
+///
+/// The macro creates a function with given name and visibility modifier,
+/// that needs to be run in order to register the resource. Calling the
+/// function more than once has no effect.
+///
+/// # Example
+///
+/// Consider this project files structure:
+/// ```text
+/// .
+/// ├── Cargo.toml
+/// ├── main.qml
+/// ├── Bar.qml
+/// └── src
+///     └── main.rs
+/// ```
+/// then the following Rust code:
+/// ```
+/// # #[macro_use] extern crate qmetaobject;
+/// # // For maintainers: this is actually tested against read files.
 /// qrc!(my_resource,
-///     "qml" {
+///     "foo" {
 ///         "main.qml",
-///         "Foo.qml" as "foo/Foo.qml",
+///         "Bar.qml" as "baz/Foo.qml",
 ///      }
 /// );
 ///
-/// //...
-/// my_resource(); // registers the resource to Qt
+/// # fn use_resource(_r: &str) {}
+/// # fn main() {
+/// // registers the resource to Qt
+/// my_resource();
+/// // do something with resources
+/// use_resource("qrc:/foo/baz/Foo.qml");
+/// # }
 /// ```
-///
 /// corresponds to the .qrc file:
-/// ```ignore
-///    <RCC>
-///        <qresource prefix="/qml">
-///            <file>main.qml</file>
-///            <file alias="foo/Foo.qml">Foo.qml</file>
-///        </qresource>
-///    </RCC>
+/// ```xml
+/// <RCC>
+///     <qresource prefix="/qml">
+///         <file>main.qml</file>
+///         <file alias="foo/Foo.qml">Foo.qml</file>
+///     </qresource>
+/// </RCC>
 /// ```
-///
-/// The paths are relative to the location in which cargo runs.
-///
-/// The macro creates a function that needs to be run in order to register the
-/// resource. Calling the function more than once has no effect.
 #[macro_export]
-macro_rules! qrc { // This is just a forwarding marco so it is documented
-    ($fn_name:ident, $($tokens:tt)* ) => {
-        qrc_internal!($fn_name, $($tokens)*);
-    }
+macro_rules! qrc {
+    // Due to a bug in Rust compiler, empty visibility modifier passed to proc macro breaks parsing
```

Comment:
have you tried jsut forwarding `$($tokens:tt)*`  ?

Also, i think it may be now possible to have the qrc macro implemented as a procedural macro, and still get the documentation in the qmetaobject crate.  is there not a new attribute so that the documentation is kept on re-exports?

Reply:
Forwarding tokens is meh. Macro signature would look less descriptive than it could be.

I have no idea about documentation re-exports, need to research on that. As far, as I understood, you want to get rid of intermediate macro call, rename and re-export qrc! macro from proc crate, but retain docs in the non-proc crate, right?

Reply:
Some info on-topic:
https://users.rust-lang.org/t/solved-is-it-possible-to-reexport-a-proc-macro/16864/

Yup, I managed re-export it from proc crate into main crate via `pub use` with new docstring and without `#[macro_blah_blah_blah]` stuff. However, docstring is not retained on re-export of the re-documented re-export.

Consider removing mod qrc or making it private, then re-export and re-document macro in lib.rs?

Reply:
yes, we can re-document the macro in lib.
However if that's a source break, maybe it can wait a bit.

Reply:
Not that I think about it, breakage is not required. Just re-export and re-document macro in lib.rs, and leave qrc.rs alone with its single internal unsafe fn. I bet, no-one used macro via `qmetaobject::qrc::qrc!` path.

Reply:
Resolved by removing the intermediate declarative macro altogether.